### PR TITLE
Feature sum

### PR DESCRIPTION
**Pull request by Jerizon**
Branch feature-sum for x-cell. Currently has functionality for adding values from columns, adding columns from the interior or to the right by default, adding rows from the interior or to the bottom by default.

### CODE REVIEW

File: client/js/array-util.js [hunk 1]
```diff
@@ -1,12 +1,24 @@
+//Returns array from fromNum to toNum that is sequential in order, both increasing and decreasing
 const getRange = function(fromNum, toNum) {
```
Comment (speschellk) [inline client/js/array-util.js:2]: Set your text editor to use 2-space tabs instead of 4-space tabs; ask me if you have any trouble. This is key for readability, especially when you checked your code into a central repo that other devs will work off of.

File: client/js/table-view.js [hunk 2]
```diff
@@ -5,84 +5,210 @@ class TableView {
...
+
+    //Adds column when called. Called by click columnButton.
+    addColumn() {
+        this.model.numCols += 1;
+        if(this.model.highlight.col === true) {
+            const colLocation = this.currentCellLocation.col;
+            for (let row = 0; row < this.model.numRows; row ++) {
+               for (let col = this.model.numCols; col > 0 ; col--) {
+                    if(col > colLocation) {
+                        const preValue = this.model.getValue({col: col - 1, row: row});
```
Comment (speschellk) [inline client/js/table-view.js]: This is the first variable name that I would have trouble knowing the function of. What is preValue? Choose a name that a dev with little other context would understand without having to comb through files or read the documentation or bug you about it.

File: client/js/test/table-view.test.js [hunk 3]
```diff
@@ -8,34 +8,109 @@ describe('table-view', () => {
         const fixturePath = './client/js/test/fixtures/sheet-container.html';
         const html = fs.readFileSync(fixturePath, 'utf8');
         document.documentElement.innerHTML = html;
-    })
+    });
+
+    describe('highlighting entire columns and rows', () => {
```
Comment (speschellk) [inline client/js/test/table-view.test.js:13]: These are good tests, but so far they test only for happy, expected paths. Add at least 2-3 others that test against edge cases. For instance, it('displays numeric total even when non-numbers are entered into column cells').

**Review by speschellk — CHANGES_REQUESTED** [review]

File: client/js/array-util.js (hunk 1 above)
Comment (Jerizon) [reply client/js/array-util.js:2]: Was able to do so in user settings of Sublime.

File: client/js/test/table-view.test.js (hunk 3 above)
Comment (Jerizon) [reply client/js/test/table-view.test.js:13]: Added edge cases that test for non-numeric values inputted in the table data, and how they would be effectively ignored by the summation.

File: client/js/table-view.js (hunk 2 above)
Comment (Jerizon) [reply client/js/table-view.js]: Changed the variable name to previousColumnStoredValue - a mouthful to be sure- to specify that it is the value that was stored in the cell and is being transferred to one row over. Also changed the variable name in the addRows() functionality to previousRowColumnStoredValue, and it serves the same purpose.

File: client/js/array-util.js (hunk 1 above)
Comment (speschellk) [reply client/js/array-util.js:2]: Excellent. For these files, you'll also need to adjust them (since they have already been created) with the bottom right button that lets you select "Tab Width: 2" and then "Convert indentation to spaces".

Comment (speschellk) [reply client/js/array-util.js:2]: Then you can commit and push up those files to this branch.

**Review by speschellk — CHANGES_REQUESTED** [review]

File: client/js/array-util.js (hunk 1 above)
Comment (Jerizon) [reply client/js/array-util.js:2]: Sounds good, I changed the indentation, hopefully it will reflect this. Thank you for your assistance, please let me know if there is anything else I can do.

Comment (speschellk) [reply client/js/array-util.js:2]: Doesn't look like it worked. Did you save each file after making those changes? git will register a change to the file if you make those changes and then save; that's how you can commit the change and push it back up here.

File: client/js/table-view.js [hunk 4]
```diff
@@ -1,88 +1,214 @@
...
+    // Renders the sum of all the columns
+    renderSumBar() {
+        const sumData = new Array(this.model.numCols);
+        for (let col = 0; col < this.model.numCols + 1; col++) {
+            sumData[col] = 0;
+            for (let row = 0; row < this.model.numRows; row++) {
+                const position = { col: col - 1, row: row };
+                const value = parseInt(this.model.getValue(position))
+                if ((typeof value === 'number') && !isNaN(value)) {
+                    sumData[col] = sumData[col] + value;
```
Comment (speschellk) [inline client/js/table-view.js:68]: This can be abbreviated as "sumData[col] += value;"

File: client/js/table-view.js [hunk 5]
```diff
@@ -1,88 +1,214 @@
...
+            }
+        }
+        removeChildren(this.sumBar);
+        sumData.map(colSum => createTD(colSum)).forEach(cs => this.sumBar.appendChild(cs));
+    }
+
+    // returns if the given location (in col, row) matches actual location  
     isCurrentCell(col, row) {
         return this.currentCellLocation.col === col &&
             this.currentCellLocation.row === row;
```
Comment (speschellk) [inline client/js/table-view.js:79]: Even better if you can line up the "this.currentCellLocation.row" just underneath its brother above. Aligning very similar lines helps a lot with readability, and looks really pleasing (something that contributes to people thinking your code is excellent).

File: client/js/table-view.js [hunk 6]
```diff
@@ -1,88 +1,214 @@
...
     attachEventHandlers() {
         this.sheetBodyEl.addEventListener('click', this.handleSheetClick.bind(this));
         this.formulaBarEl.addEventListener('keyup', this.handleFormulaBarChange.bind(this));
+        this.columnButton.addEventListener('click', this.addColumn.bind(this));
+        this.rowButton.addEventListener('click', this.addRow.bind(this));
+        this.headerRowEl.addEventListener('click', this.handleColumnClick.bind(this));
     }
 
+    //Adds column when called. Called by click columnButton.
+    addColumn() {
```
Comment (speschellk) [inline client/js/table-view.js:134]: This ended up being a large function with lots of nesting. Could you make it cleaner by using a helper function?

File: client/js/test/fixtures/sheet-container.html [hunk 7]
```diff
@@ -7,5 +7,10 @@
             <tbody>
 
             </tbody>
+            <div class='sumContainer'>
+            <tr id='sum-bar'> </tr>
```
Comment (speschellk) [inline client/js/test/fixtures/sheet-container.html:11]: Child elements should be indented 2 spaces from parent elements.

File: client/js/test/fixtures/sheet-container.html [hunk 8]
```diff
@@ -7,5 +7,10 @@
...
+            </div>
         </table>
+        <button id="addColumn" type="button">Add Column</button>
+    <button id="addRow" type="button">Add Row</button>
```
Comment (speschellk) [inline client/js/test/fixtures/sheet-container.html:15]: Should be indented another 2 spaces.

**Review by speschellk — CHANGES_REQUESTED** [review]
